qiskit_nature/operators/second_quantization/quadratic_hamiltonian.py

```diff
@@ -0,0 +1,306 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021, 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""The QuadraticHamiltonian class."""
+
+from typing import Optional, Tuple
+
+import numpy as np
+import scipy.linalg
+from qiskit.quantum_info.operators.mixins import TolerancesMixin
+from qiskit_nature.operators.second_quantization.fermionic_op import FermionicOp
+
+
+class QuadraticHamiltonian(TolerancesMixin):
+    r"""A Hamiltonian that is quadratic in the fermionic ladder operators.
+
+    A quadratic Hamiltonian is an operator of the form
+
+    .. math::
+        \sum_{p, q} M_{pq} a^\dagger_p a_q
+        + \frac12 \sum_{p, q}
+            (\Delta_{pq} a^\dagger_p a^\dagger_q + \text{h.c.})
+        + \text{constant}
+
+    where :math:`M` is a Hermitian matrix and :math:`\Delta` is an antisymmetric matrix.
+    """
+
+    def __init__(
+        self,
+        hermitian_part: Optional[np.ndarray] = None,
+        antisymmetric_part: Optional[np.ndarray] = None,
+        constant: float = 0.0,
+    ) -> None:
+        r"""Initialize a QuadraticHamiltonian.
+
+        Args:
+            hermitian_part: The matrix :math:`M` containing the
+                coefficients of the terms that conserve particle number.
+            antisymmetric_part: The matrix :math:`\Delta` containing the
+                coefficients of the terms that do not conserve particle number.
+            constant: An additive constant term.
+
+        Raises:
+            ValueError: Either a Hermitian part or antisymmetric part must be specified.
+        """
+        if hermitian_part is not None:
+            self._n_orbitals, _ = hermitian_part.shape
+        elif antisymmetric_part is not None:
+            self._n_orbitals, _ = antisymmetric_part.shape
+        else:
+            raise ValueError("Either a Hermitian part or antisymmetric part must be specified.")
```

Comment:
Would it make sense to provide a keyword argument like `num_orbitals` to handle the creation of a zero-operator?

Reply:
Done, I named it `num_modes`.